src/control/tools/SplineHandler.cpp

```diff
@@ -0,0 +1,210 @@
+                if (pointCount > 2) {
+                    // move last non dynammically changing point right
+                }
+                break;
+            }
+            case GDK_KEY_leftarrow: {
+                if (pointCount > 2) {
+                    // move last non dynammically changing point left
+                }
+                break;
+            }
+            */
+    }
+    return false;
+}
+
+auto SplineHandler::onMotionNotifyEvent(const PositionInputData& pos) -> bool {
+    if (!stroke) {
+        return false;
+    }
+
+    double zoom = xournal->getZoom();
+    int pointCount = stroke->getPointCount();
+
+    Point currentPoint = Point(pos.x / zoom, pos.y / zoom);
+    Rectangle rect = stroke->boundingRect();
+
+    if (pointCount > 0) {
+        if (!validMotion(currentPoint, stroke->getPoint(pointCount - 1))) {
+            return true;
+        }
+    }
+
+    this->redrawable->repaintRect(stroke->getX(), stroke->getY(), stroke->getElementWidth(),
+                                  stroke->getElementHeight());
+
+    drawShape(currentPoint, pos);
+
+    rect.add(stroke->boundingRect());
+    double w = stroke->getWidth();
+
+    redrawable->repaintRect(rect.x - w, rect.y - w, rect.width + 2 * w, rect.height + 2 * w);
+
+    return true;
+}
+
+void SplineHandler::onButtonReleaseEvent(const PositionInputData& pos) {
+    double zoom = xournal->getZoom();
+    if (stroke) {
+        Point currPoint = Point(pos.x / zoom, pos.y / zoom);
+        Point firstPoint = stroke->getPoint(0);
+        double dist = currPoint.lineLengthTo(firstPoint);
+        double radius = 2.0;
+        int pointCount = stroke->getPointCount();
+        if (dist < radius && pointCount > 1) {
+            stroke->setLastPoint(firstPoint);
+            finalizeSpline();
+        } else {
+            stroke->addPoint(currPoint);
+        }
+    }
+}
+
+void SplineHandler::onButtonPressEvent(const PositionInputData& pos) {
+    double zoom = xournal->getZoom();
+    if (!stroke) {
+        createStroke(Point(pos.x / zoom, pos.y / zoom));
+    }
+}
+
+void SplineHandler::onButtonDoublePressEvent(const PositionInputData& pos) { finalizeSpline(); }
+
+void SplineHandler::finalizeSpline() {
+    if (stroke == nullptr) {
+        return;
+    }
+
+    Control* control = xournal->getControl();
+
+    // This is not a valid stroke
+    if (stroke->getPointCount() < 2) {
+        g_warning("Stroke incomplete!");
+        delete stroke;
+        stroke = nullptr;
+        return;
+    }
+
+    stroke->freeUnusedPointItems();
+
+    control->getLayerController()->ensureLayerExists(page);
+
+    Layer* layer = page->getSelectedLayer();
+
+    UndoRedoHandler* undo = control->getUndoRedoHandler();
+
+    undo->addUndoAction(mem::make_unique<InsertUndoAction>(page, layer, stroke));
+
+    layer->addElement(stroke);
+    page->fireElementChanged(stroke);
+    this->redrawable->repaintRect(stroke->getX(), stroke->getY(), stroke->getElementWidth(),
```

Comment:
Use `rerenderRect` here. The difference is that `repaintRect` will draw to the view's buffer, while `rerenderRect` will draw the view's buffer to the canvas/screen (according to the documentation in `Redrawable.h`).

Reply:
Thanks for the help!